### PR TITLE
Split Vault trait

### PR DESCRIPTION
**Pull request by SanjoDeundiak**
Should be merged after #690

### CODE REVIEW

File: implementations/rust/vault/src/file/mod.rs [hunk 1]
```diff
@@ -199,42 +197,9 @@ impl Vault for FilesystemVault {
...
         self.v.ec_diffie_hellman(context, peer_public_key)
     }
+}
 
-    /// Compute Elliptic-Curve Diffie-Hellman using this secret key
-    ///
-    /// and the specified uncompressed public key and return the HKDF-SHA256
-    ///
-    /// output using the DH value as the HKDF ikm
-    fn ec_diffie_hellman_hkdf_sha256(
```
Comment (SanjoDeundiak) [inline implementations/rust/vault/src/file/mod.rs:208]: Removed as obsolete

File: implementations/rust/vault/src/file/mod.rs [hunk 2]
```diff
@@ -114,17 +120,7 @@ fn fs_write_secret(
     return Ok(());
 }
 
-impl Vault for FilesystemVault {
-    /// Generate random bytes and fill them into `data`
-    fn random(&mut self, data: &mut [u8]) -> Result<(), VaultFailError> {
```
Comment (SanjoDeundiak) [inline implementations/rust/vault/src/file/mod.rs:119]: Removed because not used

File: implementations/rust/vault/src/file/mod.rs [hunk 3]
```diff
@@ -257,30 +222,48 @@ impl Vault for FilesystemVault {
         self.v
             .aead_aes_gcm_decrypt(context, cipher_text, nonce, aad)
     }
+}
 
-    /// Close and release all resources in use by the vault
-    fn deinit(&mut self) {
```
Comment (SanjoDeundiak) [inline implementations/rust/vault/src/file/mod.rs:262]: How about move anything that was in deinit to Drop?

File: implementations/rust/vault/src/lib.rs [hunk 4]
```diff
@@ -102,33 +98,31 @@ pub trait Vault: Zeroize {
...
-        peer_public_key: &[u8],
-    ) -> Result<Box<dyn Secret>, VaultFailError>;
-    /// Compute Elliptic-Curve Diffie-Hellman using this secret key
-    /// and the specified uncompressed public key and return the HKDF-SHA256
-    /// output using the DH value as the HKDF ikm
-    fn ec_diffie_hellman_hkdf_sha256(
+}
+
+/// Trait with sign functionality
+pub trait SignerVault: Zeroize {
```
Comment (SanjoDeundiak) [inline implementations/rust/vault/src/lib.rs:104]: Decided to have 2 traits, cause sometimes one party  may need only signing or verification (e.g. signing can be on server, and client should only verify signature)

File: implementations/rust/vault/src/lib.rs [hunk 5]
```diff
@@ -102,33 +98,31 @@ pub trait Vault: Zeroize {
...
-        output_attributes: Vec<SecretAttributes>,
-    ) -> Result<Vec<Box<dyn Secret>>, VaultFailError>;
+        signature: &[u8; 64],
+        public_key: &[u8],
+        data: &[u8],
+    ) -> Result<(), VaultFailError>;
+}
+
+/// Trait with symmetric encryption
+pub trait SymmetricVault: Zeroize {
```
Comment (SanjoDeundiak) [inline implementations/rust/vault/src/lib.rs:125]: Decided not split into Encryption&Decryption traits

File: implementations/rust/vault/src/software/mod.rs [hunk 6]
```diff
@@ -650,179 +620,181 @@ mod tests {
...
-            vec![output_attributes.clone()],
-        );
-        assert!(res.is_ok());
-        let ss = res.unwrap();
-        assert_eq!(ss.len(), 1);
-        let ss = &ss[0];
-        let ss = vault.secret_export(ss).unwrap();
-        // FIXME: Check actual value
-        assert_eq!(ss.0.len(), 32);
+        // FIXME
```
Comment (SanjoDeundiak) [inline implementations/rust/vault/src/software/mod.rs:630]: Should be fixed...

File: implementations/rust/vault/src/software/mod.rs [hunk 7]
```diff
@@ -650,179 +620,181 @@ mod tests {
...
```
Comment (SanjoDeundiak) [inline implementations/rust/vault/src/software/mod.rs:716]: Should be fixed...

**Comment by SanjoDeundiak** [comment]
@mrinalwadhwa there are some elixir changes

**Review by mikelodder7 — APPROVED** [review]

**Pull request merged**